peer/peer_test.go

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2013-2013 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package peer_test
+
+import (
+	"errors"
+	"io"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/btcsuite/btcd/addrmgr"
+	"github.com/btcsuite/btcd/blockchain"
+	"github.com/btcsuite/btcd/peer"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btclog"
+)
+
+// Mocks a network connection. Implements the net.Conn interface
+type Conn struct {
+	// Local network & address for the connection
+	LNet, LAddr string
+	// Remote network & address for the connection
+	RNet, RAddr string
+	// In messages will be written to this buffer
+	In io.Writer
+	// Out messages will be read from this buffer
+	Out io.Reader
+	// If non-nil, send value to it on close
+	closec chan bool
+}
+
+func (c *Conn) Read(b []byte) (n int, err error) {
+	return c.Out.Read(b)
+}
+
+func (c *Conn) Write(b []byte) (n int, err error) {
+	return c.In.Write(b)
+}
+
+func (c Conn) LocalAddr() net.Addr {
+	return &Addr{c.LNet, c.LAddr}
+}
+
+func (c Conn) RemoteAddr() net.Addr {
+	return &Addr{c.RNet, c.RAddr}
+}
+
+func (c *Conn) Close() error {
+	select {
+	case c.closec <- true:
+	default:
+	}
+	return nil
+}
+
+// Not implemented
+func (c Conn) SetDeadline(t time.Time) error      { return nil }
+func (c Conn) SetReadDeadline(t time.Time) error  { return nil }
+func (c Conn) SetWriteDeadline(t time.Time) error { return nil }
+
+// Mocks a network address
+type Addr struct {
+	Net, Addr string
+}
+
+func (m Addr) Network() string { return m.Net }
+func (m Addr) String() string  { return m.Addr }
+
+// Pipe turns two mock connections into a full-duplex connection similar to
+// net.Pipe to allow pipe's with (fake) addresses.
+func Pipe(c1, c2 *Conn) (*Conn, *Conn) {
+	r1, w1 := io.Pipe()
+	r2, w2 := io.Pipe()
+	c1.In = w1
+	c2.Out = r1
+	c1.Out = r2
+	c2.In = w2
+
+	return c1, c2
+}
+
+// Not implemented
+func lookupFunc(host string) ([]net.IP, error) {
+	return nil, errors.New("not implemented")
+}
+
+// newestSha returns the latest known block to this peer
+func newestSha() (sha *wire.ShaHash, height int64, err error) {
+	hash := "14a0810ac680a3eb3f82edc878cea25ec41d6b790744e5daeef"
+	height = 234439
+	sha, err = wire.NewShaHashFromStr(hash)
+	return
+}
+
+func handleGetAddrMsg(p *peer.Peer, msg *wire.MsgGetAddr) {
+	addrCache := make([]*wire.NetAddress, 1)
+	addrCache[0] = p.NA()
+	err := p.PushAddrMsg(addrCache)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func handleVersionMsg(p *peer.Peer, msg *wire.MsgVersion) {
+	var zeroHash wire.ShaHash
+	locator := blockchain.BlockLocator([]*wire.ShaHash{&zeroHash})
+	p.PushGetBlocksMsg(locator, &zeroHash)
```

Comment:
I realize this is just test code, but I'd rather the test code is realistic.  Pushing a get blocks message in response to a version message goes against the protocol.